File: Python/echo_server.py
import socket
import threading

def handle_client(client_socket, addr):
    """Xử lý từng client trong thread riêng"""
    addr_str = f"{addr[0]}:{addr[1]}"
    print(f"Client kết nối: {addr_str}")
    
    try:
        while True:
            data = client_socket.recv(1024)
            if not data:
                break
            
            message = data.decode('utf-8').strip()
            print(f"Nhận: {message}")
            
            # Echo lại
            client_socket.send(data)
    except:
        pass
    finally:
        print("Client ngắt kết nối")
        client_socket.close()

def main():
    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    server.bind(('127.0.0.1', 8888))
    server.listen(5)
    
    print(" Echo Server đang chạy tại http://127.0.0.1:8888")
    print("Dùng telnet hoặc echo_client.py để test nhé!")
    
    try:
        while True:
            client_socket, addr = server.accept()
            thread = threading.Thread(target=handle_client, args=(client_socket, addr))
            thread.daemon = True
            thread.start()
    except KeyboardInterrupt:
        print("\n Server đang tắt...")
    finally:
        server.close()

if __name__ == "__main__":
    main()

File: C++/echo_server.cpp
#include <winsock2.h>
#include <iostream>
#pragma comment(lib, "ws2_32.lib")

#define PORT 8080
#define BUF_SIZE 4096

int main() {
    WSADATA wsa;
    SOCKET server, client;
    sockaddr_in addr, clientAddr;
    int clientSize = sizeof(clientAddr);
    char buffer[BUF_SIZE];

    WSAStartup(MAKEWORD(2,2), &wsa);

    server = socket(AF_INET, SOCK_STREAM, 0);

    int flag = 1, bufSize = 64 * 1024;
    setsockopt(server, IPPROTO_TCP, TCP_NODELAY, (char*)&flag, sizeof(flag));
    setsockopt(server, SOL_SOCKET, SO_SNDBUF, (char*)&bufSize, sizeof(bufSize));
    setsockopt(server, SOL_SOCKET, SO_RCVBUF, (char*)&bufSize, sizeof(bufSize));

    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = INADDR_ANY;
    addr.sin_port = htons(PORT);

    bind(server, (sockaddr*)&addr, sizeof(addr));
    listen(server, 1);

    std::cout << "Echo Server running...\n";
    client = accept(server, (sockaddr*)&clientAddr, &clientSize);

    int bytes = recv(client, buffer, BUF_SIZE, 0);
    send(client, buffer, bytes, 0); // Echo lại

    closesocket(client);
    closesocket(server);
    WSACleanup();
    return 0;
}